File: src/main/java/org/example/kalkulationsprogramm/dto/WebsiteAnalytics/AnalyticsSnapshotRequestDto.java
package org.example.kalkulationsprogramm.dto.WebsiteAnalytics;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.validation.Valid;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.PositiveOrZero;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.List;

/**
 * Eingehender Tagesschluss-Snapshot der Marketing-Website.
 * Schema-Spezifikation kommt vom Website-Team (siehe ANALYTICS_SNAPSHOT_API.md).
 * <p>
 * Unbekannte Felder werden ignoriert, damit additive Erweiterungen seitens
 * Website nicht direkt eine ERP-Anpassung erfordern (vgl. Section 5 der Spec).
 */
@Getter
@Setter
@JsonIgnoreProperties(ignoreUnknown = true)
public class AnalyticsSnapshotRequestDto {

    @NotNull
    private Integer schemaVersion;

    @NotNull
    private LocalDate snapshotDate;

    @NotNull
    private OffsetDateTime generatedAt;

    @NotNull
    @Valid
    private Totals totals;

    @PositiveOrZero
    private long visitorsToday;

    @PositiveOrZero
    private long visitorsYesterday;

    @Min(0)
    @Max(100)
    private int conversion;

    private List<FunnelStep> funnel;
    private List<TopPage> topPages;
    private List<DeviceCount> devices;
    private List<BrowserCount> browsers;
    private List<CityCount> cities;

    @Getter
    @Setter
    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class Totals {
        @PositiveOrZero
        private long visitors;
        @PositiveOrZero
        private long pageviews;
        @PositiveOrZero
        private long leadsPhone;
        @PositiveOrZero
        private long leadsMail;
        @PositiveOrZero
        private long submissions;
    }

    @Getter
    @Setter
    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class FunnelStep {
        private String name;
        private String label;
        private long count;
    }

    @Getter
    @Setter
    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class TopPage {
        private String path;
        private long count;
    }

    @Getter
    @Setter
    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class DeviceCount {
        private String device;
        private long count;
    }

    @Getter
    @Setter
    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class BrowserCount {
        private String browser;
        private long count;
    }

    @Getter
    @Setter
    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class CityCount {
        private String city;
        private String country;
        private long count;
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/dto/WebsiteAnalytics/AnalyticsSnapshotRequestDto.kt
package org.example.kalkulationsprogramm.dto.WebsiteAnalytics

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import jakarta.validation.Valid
import jakarta.validation.constraints.Max
import jakarta.validation.constraints.Min
import jakarta.validation.constraints.NotNull
import jakarta.validation.constraints.PositiveOrZero
import java.time.LocalDate
import java.time.OffsetDateTime

@JsonIgnoreProperties(ignoreUnknown = true)
data class AnalyticsSnapshotRequestDto(
    @field:NotNull var schemaVersion: Int? = null,
    @field:NotNull var snapshotDate: LocalDate? = null,
    @field:NotNull var generatedAt: OffsetDateTime? = null,
    @field:NotNull @field:Valid var totals: Totals? = null,
    @field:PositiveOrZero var visitorsToday: Long = 0,
    @field:PositiveOrZero var visitorsYesterday: Long = 0,
    @field:Min(0) @field:Max(100) var conversion: Int = 0,
    var funnel: List<FunnelStep>? = null,
    var topPages: List<TopPage>? = null,
    var devices: List<DeviceCount>? = null,
    var browsers: List<BrowserCount>? = null,
    var cities: List<CityCount>? = null,
) {
    @JsonIgnoreProperties(ignoreUnknown = true)
    data class Totals(@field:PositiveOrZero var visitors: Long = 0, @field:PositiveOrZero var pageviews: Long = 0, @field:PositiveOrZero var leadsPhone: Long = 0, @field:PositiveOrZero var leadsMail: Long = 0, @field:PositiveOrZero var submissions: Long = 0)
    @JsonIgnoreProperties(ignoreUnknown = true)
    data class FunnelStep(var name: String? = null, var label: String? = null, var count: Long = 0)
    @JsonIgnoreProperties(ignoreUnknown = true)
    data class TopPage(var path: String? = null, var count: Long = 0)
    @JsonIgnoreProperties(ignoreUnknown = true)
    data class DeviceCount(var device: String? = null, var count: Long = 0)
    @JsonIgnoreProperties(ignoreUnknown = true)
    data class BrowserCount(var browser: String? = null, var count: Long = 0)
    @JsonIgnoreProperties(ignoreUnknown = true)
    data class CityCount(var city: String? = null, var country: String? = null, var count: Long = 0)
}
